rules: no-noninteractive-tabindex, no-redundant-roles, no-noninteractive-element-to-interactive-role

ematipico opened the pull request:
Implementation of the rules:
-  [no-noninteractive-tabindex.md](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-noninteractive-tabindex.md)
- [no-redundant-roles](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-redundant-roles.md)
- [no-noninteractive-element-to-interactive-role](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-noninteractive-element-to-interactive-role.md)

Refers to #341 


This is going to be a long one unfortunately! I guess we'd need to map more elements and roles and add more tests. For now I added enough tests to cover the combination of interactive roles, interactive elements, negative tabIndex and zero tabIndex. 

I'd like some feedback in the meantime!

sebmck commented:
I have no idea how aria roles work so I'll defer review to someone else 😛

ematipico commented:
@sebmck it's all documented here: https://www.w3.org/WAI/PF/aria/roles#article

Every role has a table with all the information we need. At the moment I'm covering the most used roles and HTML elements, that's why there only some.

But once have tests that cover the edge cases, the `roles` object can be filled with another PR with more cases

ematipico commented:
The PR should be ready to be reviewed!

sebmck commented:
cc @bitpshr @VictorHom

bitpshr commented on packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveElementToInteractiveRole.ts:
```diff
@@ -0,0 +1,46 @@
+		const {node} = path;
+
+		if (isJSXElement(node)) {
+			if (hasJSXAttribute(node, "role")) {
```
Nit: These two `if` statements can be combined.

bitpshr commented on packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveElementToInteractiveRole.ts:
```diff
@@ -0,0 +1,46 @@
+					if (role) {
+						if (!isElementInteractive(name) && isRoleInteractive(role)) {
+							path.context.addNodeDiagnostic(
+								node,
```
```suggestion
								roleAttribute,
```

This will increase the specificity of the node reported in the diagnostic.

bitpshr commented on packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveTabindex.ts:
```diff
@@ -0,0 +1,123 @@
+						// not having the "widget" role means that the role is not interactive
+						if (!elementToRole.has("widget") && tabIndexValue > -1) {
+							path.context.addNodeDiagnostic(
+								node,
```
```suggestion
								tabIndexAttribute,
```

bitpshr commented on packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveTabindex.ts:
```diff
@@ -0,0 +1,123 @@
+						 */
+						if (tabIndexValue > -1) {
+							path.context.addNodeDiagnostic(
+								node,
```
```suggestion
								tabIndexAttribute,
```

bitpshr commented on packages/@romejs/compiler/lint/rules/jsx-a11y/noNoninteractiveTabindex.ts:
```diff
@@ -0,0 +1,123 @@
+						 */
+						if (!isRoleInteractive(role) && tabIndexValue > -1) {
+							path.context.addNodeDiagnostic(
+								node,
```
```suggestion
								attr,
```

bitpshr commented on packages/@romejs/diagnostics/descriptions/lint.ts:
```diff
@@ -16,6 +16,17 @@ import {buildSuggestionAdvice} from "../helpers";
 export const lint = createDiagnosticsCategory({
+	JSX_A11Y_NO_NONINTERACTIVE_ELEMENT_TO_INTERACTIVE_ROLE: (element: string) => ({
+		category: "lint/jsx-a11y/noNoninteractiveElementToInteractiveRole",
+		message: `The HTML element <emphasis>${element}</emphasis> is non-interactive and should not have an interactive role`,
```
```suggestion
		message: `The HTML element <emphasis>${element}</emphasis> is non-interactive and should not have an interactive role`.,
```

bitpshr commented on packages/@romejs/diagnostics/descriptions/lint.ts:
```diff
@@ -16,6 +16,17 @@ import {buildSuggestionAdvice} from "../helpers";
+			{
+				type: "log",
+				category: "info",
+				text: `Consider replacing <emphasis>${element}</emphasis> with "div" or "span"`,
```
```suggestion
				text: `Replace <emphasis>${element}</emphasis> with a div or a span.`,
```

bitpshr commented on packages/@romejs/diagnostics/descriptions/lint.ts:
```diff
@@ -27,6 +38,10 @@ export const lint = createDiagnosticsCategory({
 	},
+	JSX_A11Y_NO_REDUNDANT_ROLES: (role: string, element: string) => ({
+		category: "lint/jsx-a11y/noRedundantRoles",
+		message: `Using the role attribute <emphasis>${role}</emphasis> and the HTML element <emphasis>${element}</emphasis> is redundant, using the HTML element is enough.`,
```
```suggestion
		message: `Using the role attribute <emphasis>${role}</emphasis> on the <emphasis>${element}</emphasis> element is redundant.`,
```

bitpshr commented on packages/@romejs/diagnostics/descriptions/lint.ts:
```diff
@@ -38,6 +53,10 @@ export const lint = createDiagnosticsCategory({
 	}),
+	JSX_A11Y_NO_NONINTERACTIVE_TABINDEX: {
+		category: "lint/jsx-a11y/noNoninteractiveTabindex",
+		message: "Do not use <emphasis>tabIndex</emphasis> on an element that is not interactive",
```
```suggestion
		message: "Do not use <emphasis>tabIndex</emphasis> on an element that is not interactive.",
		advice: [
			{
				type: "log",
				category: "info",
				text: "Adding non-interactive elements to the keyboard navigation flow can confuse users.",
			},
		],
```

bitpshr commented on packages/@romejs/js-ast-utils/getJSXElementName.ts:
```diff
@@ -0,0 +1,20 @@
+	}
+	if (node.name.type === "JSXMemberExpression") {
+		// TODO need to understand what to do here, what's the case
+		return "";
```
The case here is a JSX tag like `<Foo.bar>foo</Foo.bar>`. We could return `"Foo.bar"` in this case, but I'm not sure how useful this actually us. I don't think this needs to be handed at all right now.

bitpshr commented on packages/@romejs/compiler/lint/rules/ariaHelpers.ts:
```diff
@@ -3,6 +3,76 @@ export type MapOfARIAPropertyDefinitions = Map<
+	| "table"
+	| "term"
+	| "toolbar";
+
```
The additions throughout this file look semantically correct. Do you think it would be worth splitting this file into smaller modules, and putting them in an `rules/aria/` folder? This file is getting large and I don't think we've covered all roles yet.

ematipico replied:
Yes, in my opinion it's worth splitting up this file in different different files under a folder, I already do that it will get bigger and bigger! I can do that in another PR.

ematipico commented:
@bitpshr I addressed all your suggestions!

I'd prefer to implement/add more test cases on a second PR if possible! This PR is already big enough and it adds lot of utilities and changes that could be used for other rules.
